src/main/java/sirius/db/mixing/SmartQuery.java

```diff
@@ -170,11 +172,86 @@ public EntityDescriptor getEntityDescriptor() {
      * @return the query itself for fluent method calls
      */
     public SmartQuery<E> fields(Column... fields) {
-        this.fields = Arrays.asList(fields);
+        this.fields = new ArrayList<>(Arrays.asList(fields));
+        checkJoinColumnsPresent();
+
         return this;
     }
 
     /**
+     * Aggregates all columns which must be present to fetch all join-columns so that the {@link EntityRef}
+     * implementation works correctly.
+     *
+     * @return a map of columns which need to be present
+     */
+    private HashMap<String, Column> aggregateRequiredJoinColumns() {
+        HashMap<String, Column> requiredJoinFields = new HashMap<>();
+
+        this.fields.forEach(field -> {
+            Monoflop mf = Monoflop.create();
+            StringBuilder fieldName = new StringBuilder();
+            Column currentField = field.getParent();
+
+            while (currentField != null) {
+                if (mf.successiveCall()) {
+                    fieldName.append(".");
+                }
+
+                fieldName.append(currentField.getName());
+
+                if (!requiredJoinFields.containsKey(fieldName.toString())) {
+                    requiredJoinFields.put(fieldName.toString(), currentField);
+                }
+
+                currentField = currentField.getParent();
+            }
+        });
+
+        return requiredJoinFields;
+    }
+
+    /**
+     * Aggregates all columns which where explicitly set and should be fetched using {@link #fields(Column...)}.
+     *
+     * @return a set of columns which are about to be fetched
+     */
+    private Set<String> aggregatePresentColumns() {
+        Set<String> presentColumnPaths = new HashSet<>();
+
+        this.fields.forEach(field -> {
+            Monoflop mf = Monoflop.create();
+            StringBuilder fieldName = new StringBuilder();
+            Column currentField = field;
+
+            while (currentField != null) {
+                if (mf.successiveCall()) {
+                    fieldName.append(".");
+                }
+                fieldName.append(currentField.getName());
+                currentField = currentField.getParent();
+            }
+
+            presentColumnPaths.add(fieldName.toString());
+        });
+
+        return presentColumnPaths;
+    }
+
+    /**
+     * Synchonizes the columns which where explicitly set using {@link #fields(Column...)} with the columns that are
+     * required for join-fetches. If columns are missing, they are automatically added.
+     */
+    private void checkJoinColumnsPresent() {
+        Set<String> presentColumnPaths = aggregatePresentColumns();
```

Comment:
couldn't we build a single stream to do all this using map instead of severyl intermediate data structures?